### PR TITLE
executor: use file descriptor for process stdout/stderr

### LINKED ISSUE #5472: exec driver: wait detection regression
### Nomad version
Nomad v0.9.0-rc1 (7c00ab4f3f37cfd1e258b38fd2ad99e7bc23e4c3)

### Operating system and Environment details
Ubuntu 16.04

### Issue
In 0.9.0-rc1, when an exec task spawns multiple processes, exec driver blocks until all processes die.  Nomad 0.8.7 blocks until the init process dies only.

### Reproduction steps

1. Spin up nomad in dev mode: `nomad agent -dev`
2. Submit the attach job by running `nomad job run example.nomad`
2. Monitor alloc status with `nomad alloc status ...`

Also edit the sleep values in foreground and background task, and monitor events; also spin up task with long sleep values and attempt to kill processes manually with KILL.

Nomad 0.9.0-rc1 in all cases seem to terminate only when both sleep processes terminate/killed, while 0.8.7 always stops when the foreground sleep process terminates/killed.

### On Nomad 0.9.0-rc1 Behavior

Note that the taks is marked as running for ~30 seconds as measured by `Started` and `Terminated` event time:

```
vagrant@linux:~/go/src/github.com/hashicorp$ nomad version
Nomad v0.9.0-rc1 (7c00ab4f3f37cfd1e258b38fd2ad99e7bc23e4c3)
vagrant@linux:~/go/src/github.com/hashicorp$ nomad job run ./example.nomad
==> Monitoring evaluation "1278b3cf"
    Evaluation triggered by job "example"
    Allocation "59688ba9" created: node "6d2c9d26", group "cache"
    Evaluation status changed: "pending" -> "complete"
==> Evaluation "1278b3cf" finished with status "complete"
vagrant@linux:~/go/src/github.com/hashicorp$ sleep 35; nomad alloc status 59688ba9
ID                  = 59688ba9
Eval ID             = 1278b3cf
Name                = example.cache[0]
Node ID             = 6d2c9d26
Job ID              = example
Job Version         = 0
Client Status       = pending
Client Description  = No tasks have started
Desired Status      = run
Desired Description = <none>
Created             = 48s ago
Modified            = 16s ago

Task "redis" is "pending"
Task Resources
CPU        Memory           Disk     Addresses
0/500 MHz  200 KiB/256 MiB  300 MiB

Task Events:
Started At     = 2019-03-26T19:48:02Z
Finished At    = N/A
Total Restarts = 1
Last Restart   = 2019-03-26T19:48:32Z

Recent Events:
Time                  Type        Description
2019-03-26T19:48:32Z  Restarting  Task restarting in 17.51859489s
2019-03-26T19:48:32Z  Terminated  Exit Code: 0
2019-03-26T19:48:02Z  Started     Task started by client
2019-03-26T19:48:00Z  Task Setup  Building Task Directory
2019-03-26T19:48:00Z  Received    Task received by client
```

### On Nomad 0.8.7 status

Note for 0.8.7, the same job result in a task is marked as terminated after ~7 seconds:

```
vagrant@linux:~/go/src/github.com/hashicorp$ nomad version
Nomad v0.8.7 (21a2d93eecf018ad2209a5eab6aae6c359267933+CHANGES)
vagrant@linux:~/go/src/github.com/hashicorp$ nomad job run ./example.nomad
==> Monitoring evaluation "0a5a3e89"
    Evaluation triggered by job "example"
    Allocation "00a3396b" created: node "93f566cc", group "cache"
    Evaluation status changed: "pending" -> "complete"
==> Evaluation "0a5a3e89" finished with status "complete"
vagrant@linux:~/go/src/github.com/hashicorp$ sleep 10; nomad alloc status 00a3396b
ID                  = 00a3396b
Eval ID             = 0a5a3e89
Name                = example.cache[0]
Node ID             = 93f566cc
Job ID              = example
Job Version         = 0
Client Status       = pending
Client Description  = <none>
Desired Status      = run
Desired Description = <none>
Created             = 21s ago
Modified            = 12s ago

Task "redis" is "pending"
Task Resources
CPU      Memory   Disk     IOPS  Addresses
500 MHz  256 MiB  300 MiB  0

Task Events:
Started At     = 2019-03-26T19:50:57Z
Finished At    = N/A
Total Restarts = 1
Last Restart   = 2019-03-26T19:51:03Z

Recent Events:
Time                  Type        Description
2019-03-26T19:51:03Z  Restarting  Task restarting in 17.744764288s
2019-03-26T19:51:03Z  Terminated  Exit Code: 0
2019-03-26T19:50:56Z  Started     Task started by client
2019-03-26T19:50:55Z  Task Setup  Building Task Directory
2019-03-26T19:50:55Z  Received    Task received by client
```

### Job file (if appropriate)
```
job "example" {
  datacenters = ["dc1"]

  group "cache" {
    task "redis" {
      driver = "exec"

      config {
        command = "/bin/bash"
        #args = ["-c", "(sleep 200 > /dev/null & ) ; exec sleep 500"]
        args = ["-c", "(sleep 30 > /dev/null & ) ; exec sleep 5"]
      }

      resources {
        cpu    = 500
        memory = 256
      }

    }
  }
}
```

### PR DESCRIPTION
**Pull request by nickethier**
This PR introduced a new package `procio` which handles process io. In order to fix #5472 a file descriptor should be given to the command instead of an `io.Writer` interface. This way `Command.Wait` will return when the parent process dies instead of waiting for all children to terminate.

### CODE REVIEW

File: drivers/shared/procio/io.go [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+	Stderr() io.ReadCloser
+	Set(SetWritersCB)
+}
+
+type IOType string
+
+const (
+	IOTypeEmpty  IOType = ""
+	IOTypeFIFO   IOType = "fifo"
+	IOTypeBuffer IOType = "buffer"
```
Comment (notnoop) [inline drivers/shared/procio/io.go]: Should we add `null` explicitly too?

File: drivers/shared/procio/io.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+	return s.IOType == "" && s.Stdout == "" && s.Stderr == ""
+}
+
+// ProcessIO wraps an IO interface and and sends output to the configured
+// destination
+type ProcessIO struct {
+	io IO
+
+	copyLoop bool
+	stdio    Stdio
```
Comment (notnoop) [inline drivers/shared/procio/io.go]: `IO` is overused in this package, making it hard to follow the distinctions, e.g. `io` vs `stdio`.  Maybe we can name them by usage? `internalIO` vs `configIO`?  Or a more descriptive comment?

File: drivers/shared/executor/executor_test.go [hunk 3]
```diff
@@ -139,8 +129,8 @@ func TestExecutor_Start_Wait(pt *testing.T) {
 
 			expected := "hello world"
 			tu.WaitForResult(func() (bool, error) {
-				outWriter, _ := execCmd.GetWriters()
-				output := outWriter.(*bufferCloser).String()
+				out, _ := executor.(*LibcontainerExecutor).pio.IO().(*procio.BufferIO).Buffers()
```
Comment (notnoop) [inline drivers/shared/executor/executor_test.go]:
```suggestion
				out, _ := executor.(procioGetter).getProcIO().IO().(*procio.BufferIO).Buffers()
```

File: drivers/shared/executor/executor.go [hunk 4]
```diff
@@ -275,17 +233,16 @@ func (e *UniversalExecutor) Launch(command *ExecCommand) (*ProcessState, error)
...
-	stdout, err := e.commandCfg.Stdout()
-	if err != nil {
-		return nil, err
-	}
-	stderr, err := e.commandCfg.Stderr()
+	pio, err := procio.NewProcessIO(e.commandCfg.stdio(),
+		func(out, err io.WriteCloser) {
+			e.childCmd.Stdout = out
+			e.childCmd.Stderr = err
+		})
```
Comment (notnoop) [inline drivers/shared/executor/executor.go]: I'd suggest using the getters instead of adding callbacks and a callback type.  The following seems just as expressive without callback overhead:

```go
pio, err := procio.NewProcessIO(e.commandCfg.stdio())
if err != nil { return nil, err }
e.childCmd.Stdout, e.childCmd.Stderr = pio.IO().Stdout(), pio.IO().Stderr()
```

File: drivers/shared/executor/executor.go [hunk 5]
```diff
@@ -41,6 +41,12 @@ var (
 	ExecutorBasicMeasuredCpuStats = []string{"System Mode", "User Mode", "Percent"}
 )
 
+// procioGetter is only used for testing and allows tests to access process
+// io without needing to perform type assertion on the executor implementation
+type procioGetter interface {
+	getProcIO() *procio.ProcessIO
```
Comment (notnoop) [inline drivers/shared/executor/executor.go]: How do you feel about this being part of `Executor` interface directly?  It avoids the cast made in tests.

File: drivers/shared/executor/executor_linux.go [hunk 6]
```diff
@@ -355,6 +366,9 @@ func (l *LibcontainerExecutor) Shutdown(signal string, grace time.Duration) erro
 	case <-time.After(time.Second * 15):
 		return fmt.Errorf("process failed to exit after 15 seconds")
 	}
+
+	l.iowg.Wait()
```
Comment (notnoop) [inline drivers/shared/executor/executor_linux.go]: Should this be covered by the 15 second timeout above?  Also, in the universal executor case, it'd be nice to have a grace period to copy output before killing processes.

File: drivers/shared/executor/executor_linux.go [hunk 7]
```diff
@@ -164,23 +174,24 @@ func (l *LibcontainerExecutor) Launch(command *ExecCommand) (*ProcessState, erro
...
 
-	// the task process will be started by the container
-	process := &libcontainer.Process{
-		Args:   combined,
-		Env:    command.Env,
-		Stdout: stdout,
-		Stderr: stderr,
-		Init:   true,
-	}
+	l.pio = pio
```
Comment (notnoop) [inline drivers/shared/executor/executor_linux.go]: Do we need to call `l.pio.Copy(...)` somewhere here?

File: drivers/shared/executor/executor.go (hunk 5 above)
Comment (nickethier) [reply drivers/shared/executor/executor.go]: It just means that the grpc client has to implement it, which is fine just odd.

File: drivers/shared/procio/io.go [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+	fw, err := fifo.Open(path)
+	if err != nil {
+		return err
+	}
+
+	wg.Add(1)
+	go func() {
+		p := bufPool.Get().(*[]byte)
+		defer bufPool.Put(p)
+		io.CopyBuffer(fw, r, *p)
```
Comment (schmichael) [inline drivers/shared/procio/io.go]: Is there anything to be done with errors from copybuffer? They could be stored in fields on ProcessIO and exposed via `StdoutErr() error` methods, but I'm not sure there's any point?

File: drivers/shared/procio/io.go [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+		return p.io.Close()
+	}
+	return nil
+}
+
+func (p *ProcessIO) IO() IO {
+	return p.io
+}
+
+func (p *ProcessIO) Copy(wg *sync.WaitGroup) error {
```
Comment (schmichael) [inline drivers/shared/procio/io.go]: Comment.

Passing around a WaitGroup seems like a fragile API. `Copy() (<-chan struct{}, error)` would take a little more work in ProcessIO but allow callers to block on Copy in a select and exit after a timeout.

File: drivers/shared/procio/io.go (hunk 2 above)
Comment (schmichael) [reply drivers/shared/procio/io.go]: `sink` and `source` may be good variable names to illustrate which direction the data flows.

File: drivers/shared/procio/pipe.go [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+	)
+
+	// cleanup in case of an error
+	defer func() {
+		if err != nil {
+			for _, p := range pipes {
+				p.Close()
+			}
+		}
+	}()
```
Comment (schmichael) [inline drivers/shared/procio/pipe.go]: This is far more work than just calling `stdout.Close` on line 61.

File: drivers/shared/procio/pipe.go [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+		i.out,
+		i.err,
+	} {
+		if p != nil {
+			// capture first close error if occurs
+			if closeErr := p.Close(); err == nil {
+				err = closeErr
+			}
+		}
+	}
```
Comment (schmichael) [inline drivers/shared/procio/pipe.go]: This is far more work than just:

```go
outErr := i.out.Close()
if err := i.err.Close(); outErr == nil {
  outErr = err
}
return outErr
```

File: drivers/shared/procio/pipe.go [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+	err := p.w.Close()
+	if rerr := p.r.Close(); err == nil {
+		err = rerr
+	}
+	return err
+}
+
+type pipeIO struct {
+	out *pipe
+	err *pipe
```
Comment (schmichael) [inline drivers/shared/procio/pipe.go]: I'm never a fan of a variable named `err` that isn't an `error`.

File: drivers/shared/procio/pipe_unix.go [hunk 13]
```diff
@@ -0,0 +1,3 @@
+// +build !windows
+
+package procio
```
Comment (schmichael) [inline drivers/shared/procio/pipe_unix.go]: ?

File: drivers/shared/executor/executor.go (hunk 4 above)
Comment (schmichael) [reply drivers/shared/executor/executor.go]: Also, setting `e.childCmd.Std{out,err}` to the correct type (`*os.File`) is the critical part of this diff for getting the proper `childCmd.Wait` behavior, right? If so let's call that out with a big huge comment.

File: drivers/shared/executor/executor_linux.go (hunk 6 above)
Comment (schmichael) [reply drivers/shared/executor/executor_linux.go]: I'm curious about the grace period to copy output. My understanding is that using a pipe, it will get closed when the child process exits, so all logs will already be inflight through the pipe and block Wait until they're copied?

File: drivers/shared/procio/io.go [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+		return err
+	}
+
+	wg.Add(1)
+	go func() {
+		p := bufPool.Get().(*[]byte)
+		defer bufPool.Put(p)
+		io.CopyBuffer(fw, r, *p)
+		wg.Done()
+		fw.Close()
```
Comment (schmichael) [inline drivers/shared/procio/io.go]: Why do you call Done before Close? It seems like that could cause a race with Wait unblocking and exiting before everything is flushed to the fifo?
